server/src/main/java/org/elasticsearch/common/util/concurrent/ListenableFuture.java

```diff
@@ -9,122 +9,32 @@
-     * thread.
-     */
-    public void addListener(ActionListener<V> listener) {
-        addListener(listener, EsExecutors.DIRECT_EXECUTOR_SERVICE, null);
-    }
-
-    /**
-     * Adds a listener to this future. If the future has not yet completed, the listener will be notified of a response or exception in a
-     * runnable submitted to the {@link Executor} provided. If the future has completed, the listener will be notified immediately without
-     * forking to a different thread.
-     *
-     * It will restore the provided {@link ThreadContext} (if not null) when completing the listener.
-     */
-    public void addListener(ActionListener<V> listener, Executor executor, @Nullable ThreadContext threadContext) {
-        if (done || addListenerIfIncomplete(listener, executor, threadContext) == false) {
-            // run the callback directly, we don't hold the lock and don't need to fork!
-            notifyListener(listener);
-        }
-    }
-
-    @Override
-    protected void done(boolean ignored) {
-        final var existingListeners = acquireExistingListeners();
-        if (existingListeners != null) {
-            for (final var listener : existingListeners) {
-                notifyListener(listener);
-            }
-        }
-    }
-
-    private synchronized boolean addListenerIfIncomplete(ActionListener<V> listener, Executor executor, ThreadContext threadContext) {
-        // check done under lock since it could have been modified; also protect modifications to the list under lock
-        if (done) {
-            return false;
-        }
-        if (threadContext != null) {
-            listener = ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
-        }
-        if (executor != EsExecutors.DIRECT_EXECUTOR_SERVICE) {
-            listener = new ThreadedActionListener<>(executor, listener);
-        }
-        if (listeners == null) {
-            listeners = new ArrayList<>();
-        }
-        listeners.add(listener);
-        return true;
-    }
-
-    private synchronized List<ActionListener<V>> acquireExistingListeners() {
+public final class ListenableFuture<T> extends FanOutListener<T> {
+    public T result() {
         try {
-            done = true;
-            return listeners;
-        } finally {
-            listeners = null;
-        }
-    }
-
-    private void notifyListener(ActionListener<V> listener) {
-        assert done;
-        // call non-blocking result() as we could be on a network or scheduler thread which we must not block
-        ActionListener.completeWith(listener, this::result);
-    }
-
-    @Override
-    public void onResponse(V v) {
-        final boolean set = set(v);
-        if (set == false) {
-            assert false;
-            throw new IllegalStateException("did not set value, value or exception already set?");
```

Comment:
NB we relax this check for double-completion here (it doesn't matter, and the assertions show that nobody was relying on it anyway)